In [1]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import plotly.express as px

In [12]:

import pandas as pd
import plotly.express as px
import numpy as np
import plotly.graph_objects as go
import matplotlib.pylab as plt
import seaborn as sns

from IPython.display import Image

### Data Loading

In [2]:
sales = pd.read_csv('sales_train_evaluation.csv')
calendar = pd.read_csv('calendar.csv')
prices = pd.read_csv('sell_prices.csv')

### Basic information about data 

In [3]:
sales.head(5)

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_evaluation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_evaluation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_evaluation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_evaluation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_evaluation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id  d_1  d_2  d_3  d_4  ...  d_1932  d_1933  d_1934  d_1935  d_1936  \
0       CA    0    0    0    0  ...       2       4       0       0       0   
1       CA    0    0    0    0  ...       0       1       2       1       1   
2       CA    0    0    0    0  ...       1       0       2       0       0   
3       CA    0    0    0    0  ...       1       1       0       4       0   
4       CA    0    0    0    0  ...       0       0       0       2       1   

   d_1937  d_1938  d_1939  d_1940  d_1941  
0       0       3       3       0       1  
1       0       0       0       0       0  
2       0       2       3       0       1  
3       1       3       0       2       6  
4       0       0       2       1       0  

[5 rows x 1947 columns]

In [42]:
sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30490 entries, 0 to 30489
Columns: 1947 entries, id to d_1941
dtypes: int64(1941), object(6)
memory usage: 452.9+ MB


### Observation:
1. sales file contain  30490 rows and 1947 columns
2. this file contian all inforamtion about sale (sales on perticular day) and cantain product id, state id,categori id, deperament id. 
2. file size is 452.9+ MB

In [4]:
calendar.head()

date  wm_yr_wk    weekday  wday  month  year    d event_name_1  \
0  2011-01-29     11101   Saturday     1      1  2011  d_1          NaN   
1  2011-01-30     11101     Sunday     2      1  2011  d_2          NaN   
2  2011-01-31     11101     Monday     3      1  2011  d_3          NaN   
3  2011-02-01     11101    Tuesday     4      2  2011  d_4          NaN   
4  2011-02-02     11101  Wednesday     5      2  2011  d_5          NaN   

  event_type_1 event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0          NaN          NaN          NaN        0        0        0  
1          NaN          NaN          NaN        0        0        0  
2          NaN          NaN          NaN        0        0        0  
3          NaN          NaN          NaN        1        1        0  
4          NaN          NaN          NaN        1        0        1

In [5]:
# here i am converting NaN value to Not_event
calendar.fillna('No_event',inplace=True)
calendar.head()

date  wm_yr_wk    weekday  wday  month  year    d event_name_1  \
0  2011-01-29     11101   Saturday     1      1  2011  d_1     No_event   
1  2011-01-30     11101     Sunday     2      1  2011  d_2     No_event   
2  2011-01-31     11101     Monday     3      1  2011  d_3     No_event   
3  2011-02-01     11101    Tuesday     4      2  2011  d_4     No_event   
4  2011-02-02     11101  Wednesday     5      2  2011  d_5     No_event   

  event_type_1 event_name_2 event_type_2  snap_CA  snap_TX  snap_WI  
0     No_event     No_event     No_event        0        0        0  
1     No_event     No_event     No_event        0        0        0  
2     No_event     No_event     No_event        0        0        0  
3     No_event     No_event     No_event        1        1        0  
4     No_event     No_event     No_event        1        0        1

In [6]:
calendar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1969 entries, 0 to 1968
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   date          1969 non-null   object
 1   wm_yr_wk      1969 non-null   int64 
 2   weekday       1969 non-null   object
 3   wday          1969 non-null   int64 
 4   month         1969 non-null   int64 
 5   year          1969 non-null   int64 
 6   d             1969 non-null   object
 7   event_name_1  1969 non-null   object
 8   event_type_1  1969 non-null   object
 9   event_name_2  1969 non-null   object
 10  event_type_2  1969 non-null   object
 11  snap_CA       1969 non-null   int64 
 12  snap_TX       1969 non-null   int64 
 13  snap_WI       1969 non-null   int64 
dtypes: int64(7), object(7)
memory usage: 215.5+ KB


### Observation :
1. calendar file contain  1969 rows and 14 columns.
2. file contain all infoprmation about events, day month, year.
3. file size is 215.5+ KB

In [46]:
prices.head()

store_id        item_id  wm_yr_wk  sell_price
0     CA_1  HOBBIES_1_001     11325        9.58
1     CA_1  HOBBIES_1_001     11326        9.58
2     CA_1  HOBBIES_1_001     11327        8.26
3     CA_1  HOBBIES_1_001     11328        8.26
4     CA_1  HOBBIES_1_001     11329        8.26

In [47]:
prices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6841121 entries, 0 to 6841120
Data columns (total 4 columns):
 #   Column      Dtype  
---  ------      -----  
 0   store_id    object 
 1   item_id     object 
 2   wm_yr_wk    int64  
 3   sell_price  float64
dtypes: float64(1), int64(1), object(2)
memory usage: 208.8+ MB


In [7]:
pd.isnull(prices['sell_price']).value_counts()

False    6841121
Name: sell_price, dtype: int64

### Observation :
1. calendar file contain 6841121 rows and 4 columns.
2. file contain all information about average sale price of perticular product in single week.
3. file size is 215.5+ KB

In [8]:
#1. converting the wide data format to long data format and merged all the datasets
# ata sales, calendar, and sell prices to a single dataframe using common “d_” key (day number).
df_sales=pd.melt(sales, id_vars=['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id'], var_name='d', value_name='sold')

In [9]:
df_sales.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_evaluation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_evaluation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_evaluation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_evaluation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_evaluation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id    d  sold  
0       CA  d_1     0  
1       CA  d_1     0  
2       CA  d_1     0  
3       CA  d_1     0  
4       CA  d_1     0

###  Data gathered in single DataFrame 

In [10]:
df_sales.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_evaluation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_evaluation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_evaluation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_evaluation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_evaluation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id    d  sold  
0       CA  d_1     0  
1       CA  d_1     0  
2       CA  d_1     0  
3       CA  d_1     0  
4       CA  d_1     0

### 1. some info about  "wm_yr_wk" from prices dataframe:
1. wm_yr_wk represent week number.
2. in prices dataframe there is store_id,item_id,wm_yr_wk and sell_price. in that the sell_price mean that product average price on that perticular week.

### 2. strategy to assign price to each product. and explanation about this code line " pd.merge(df_final, prices, on=['store_id','item_id','wm_yr_wk'], how='left') "
3. so aim is to gather the all dataframe in single  dataframe right. so when it comes to prices dataframe we dont have explicit prise for each day insted we have price for each week.
4. for some product, price could be diferent for different state, store,period.
5. challenge is to assign right price to particular product considering state, store and  period.
**note: store id contain information about store id as well state id.**
6. so first we have to merge sales_calender dataframe with price based on state_is.
    note: store id contain information about store id as well state id. then  on store_id then item_id then finally on wm_yr_wk (week_id).  

In [11]:
df_final= pd.merge(df_sales, calendar, on='d', how='left')

In [12]:
df_final= pd.merge(df_final, prices, on=['store_id','item_id','wm_yr_wk'], how='left') 

In [13]:
df_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 59181090 entries, 0 to 59181089
Data columns (total 22 columns):
 #   Column        Dtype  
---  ------        -----  
 0   id            object 
 1   item_id       object 
 2   dept_id       object 
 3   cat_id        object 
 4   store_id      object 
 5   state_id      object 
 6   d             object 
 7   sold          int64  
 8   date          object 
 9   wm_yr_wk      int64  
 10  weekday       object 
 11  wday          int64  
 12  month         int64  
 13  year          int64  
 14  event_name_1  object 
 15  event_type_1  object 
 16  event_name_2  object 
 17  event_type_2  object 
 18  snap_CA       int64  
 19  snap_TX       int64  
 20  snap_WI       int64  
 21  sell_price    float64
dtypes: float64(1), int64(8), object(13)
memory usage: 10.1+ GB


In [14]:
df_final['item_id'].unique()

array(['HOBBIES_1_001', 'HOBBIES_1_002', 'HOBBIES_1_003', ...,
       'FOODS_3_825', 'FOODS_3_826', 'FOODS_3_827'], dtype=object)

In [15]:
# df_final.info()
pd.isnull(df_final["sell_price"]).value_counts()

False    46881677
True     12299413
Name: sell_price, dtype: int64

### Observation:
1. merging the two dataframe based on id then subject_id.
2. if you look at row 3 thats only row where id is same then subject id is same.

# 1. EDA

## 1.1 Distribution of product in USA 

In [16]:
# Grouping data based on ('state_id','store_id','cat_id','dept_id')
gp = sales.groupby(['state_id','store_id','cat_id','dept_id'],as_index=False)

In [17]:
# counting item in each group, and droping all NaN values
gp = gp['item_id'].count().dropna()
gp

state_id store_id     cat_id      dept_id  item_id
0        CA     CA_1      FOODS      FOODS_1      216
1        CA     CA_1      FOODS      FOODS_2      398
2        CA     CA_1      FOODS      FOODS_3      823
3        CA     CA_1    HOBBIES    HOBBIES_1      416
4        CA     CA_1    HOBBIES    HOBBIES_2      149
..      ...      ...        ...          ...      ...
65       WI     WI_3      FOODS      FOODS_3      823
66       WI     WI_3    HOBBIES    HOBBIES_1      416
67       WI     WI_3    HOBBIES    HOBBIES_2      149
68       WI     WI_3  HOUSEHOLD  HOUSEHOLD_1      532
69       WI     WI_3  HOUSEHOLD  HOUSEHOLD_2      515

[70 rows x 5 columns]

In [18]:
# now we are adding one more column to gp with value USA
gp['USA'] = 'USA'
gp

state_id store_id     cat_id      dept_id  item_id  USA
0        CA     CA_1      FOODS      FOODS_1      216  USA
1        CA     CA_1      FOODS      FOODS_2      398  USA
2        CA     CA_1      FOODS      FOODS_3      823  USA
3        CA     CA_1    HOBBIES    HOBBIES_1      416  USA
4        CA     CA_1    HOBBIES    HOBBIES_2      149  USA
..      ...      ...        ...          ...      ...  ...
65       WI     WI_3      FOODS      FOODS_3      823  USA
66       WI     WI_3    HOBBIES    HOBBIES_1      416  USA
67       WI     WI_3    HOBBIES    HOBBIES_2      149  USA
68       WI     WI_3  HOUSEHOLD  HOUSEHOLD_1      532  USA
69       WI     WI_3  HOUSEHOLD  HOUSEHOLD_2      515  USA

[70 rows x 6 columns]

In [19]:
# https://medium.com/thecyphy/m5-forecasting-accuracy-af6c45fb7d58
gp = sales.groupby(['state_id','store_id','cat_id','dept_id'],as_index=False)
gp = gp['item_id'].count().dropna()
gp['USA'] = 'USA'

fig = px.treemap(gp, path=['USA', 'state_id', 'store_id', 'cat_id', 'dept_id'], values='item_id',
                  color='item_id',
                  color_continuous_scale = 'RdBu',
                  title='Walmart: Distribution of items')
            
fig.update_layout(treemapcolorway = ["pink", "lightgray"])
fig.show()

### 1.2.1 mean sales of product in USA 

In [20]:
# here i am grouping data by days and takeing mean 
sales_mean=df_final.groupby(['d'],as_index=False,sort=False).agg(mean_sold=('sold',np.mean))
sales_mean

d  mean_sold
0        d_1   1.070220
1        d_2   1.041292
2        d_3   0.780026
3        d_4   0.833454
4        d_5   0.627944
...      ...        ...
1936  d_1937   1.216661
1937  d_1938   1.212299
1938  d_1939   1.395605
1939  d_1940   1.689669
1940  d_1941   1.782158

[1941 rows x 2 columns]

In [26]:
#https://plotly.com/python/time-series/
# Time Series With Range Slider
import plotly.express as px
import pandas as pd
import IPython
fig = px.line(sales_mean, x='d', y='mean_sold', title='USA VS mean sale')
fig.update_xaxes(rangeslider_visible=True)
fig.show()

### Observations:
1. every year in USA sales increases.
2. if you look at a to the plot closely every year sales follow some pattern or you can found seasonal effect.
3. every monday sales start to fall and after 3 day again sales start to go up.
4. these are some days where sale was zero.days:d_331,d_697,d_1062,d_1427,d_1792.i am going to find this when i will work with month ,day, year data.
5. if you observe plot closely you can found some trend which is going upword year by year.

* after digging deep into data then i relaised that on every Christmas sales is almost zero. only product belong to food category having some sale in Christmas.

In [22]:
data=df_final[(df_final['d']=='d_331')]
data.head()

id        item_id    dept_id   cat_id  \
10061700  HOBBIES_1_001_CA_1_evaluation  HOBBIES_1_001  HOBBIES_1  HOBBIES   
10061701  HOBBIES_1_002_CA_1_evaluation  HOBBIES_1_002  HOBBIES_1  HOBBIES   
10061702  HOBBIES_1_003_CA_1_evaluation  HOBBIES_1_003  HOBBIES_1  HOBBIES   
10061703  HOBBIES_1_004_CA_1_evaluation  HOBBIES_1_004  HOBBIES_1  HOBBIES   
10061704  HOBBIES_1_005_CA_1_evaluation  HOBBIES_1_005  HOBBIES_1  HOBBIES   

         store_id state_id      d  sold        date  wm_yr_wk  ... month  \
10061700     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061701     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061702     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061703     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061704     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   

          year  event_name_1  event_type_1 event_name_2 event_type_2 snap_CA  \
10061700  2011     Christmas      National     No_event     No_event       0   
10061701  2011     Christmas      National     No_event     No_event       0   
10061702  2011     Christmas      National     No_event     No_event       0   
10061703  2011     Christmas      National     No_event     No_event       0   
10061704  2011     Christmas      National     No_event     No_event       0   

         snap_TX  snap_WI  sell_price  
10061700       0        0         NaN  
10061701       0        0        3.97  
10061702       0        0         NaN  
10061703       0        0        4.34  
10061704       0        0        2.48  

[5 rows x 22 columns]

In [28]:
data=df_final[df_final['event_name_1']=='Christmas']
data.head()

id        item_id    dept_id   cat_id  \
10061700  HOBBIES_1_001_CA_1_evaluation  HOBBIES_1_001  HOBBIES_1  HOBBIES   
10061701  HOBBIES_1_002_CA_1_evaluation  HOBBIES_1_002  HOBBIES_1  HOBBIES   
10061702  HOBBIES_1_003_CA_1_evaluation  HOBBIES_1_003  HOBBIES_1  HOBBIES   
10061703  HOBBIES_1_004_CA_1_evaluation  HOBBIES_1_004  HOBBIES_1  HOBBIES   
10061704  HOBBIES_1_005_CA_1_evaluation  HOBBIES_1_005  HOBBIES_1  HOBBIES   

         store_id state_id      d  sold        date  wm_yr_wk  ... month  \
10061700     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061701     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061702     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061703     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   
10061704     CA_1       CA  d_331     0  2011-12-25     11148  ...    12   

          year  event_name_1  event_type_1 event_name_2 event_type_2 snap_CA  \
10061700  2011     Christmas      National     No_event     No_event       0   
10061701  2011     Christmas      National     No_event     No_event       0   
10061702  2011     Christmas      National     No_event     No_event       0   
10061703  2011     Christmas      National     No_event     No_event       0   
10061704  2011     Christmas      National     No_event     No_event       0   

         snap_TX  snap_WI  sell_price  
10061700       0        0         NaN  
10061701       0        0        3.97  
10061702       0        0         NaN  
10061703       0        0        4.34  
10061704       0        0        2.48  

[5 rows x 22 columns]

In [29]:
data=data.groupby(['store_id','cat_id'],as_index=False,sort=False).agg(Total_sale=('sold',np.sum))
data

store_id     cat_id  Total_sale
0      CA_1    HOBBIES           0
1      CA_1  HOUSEHOLD           0
2      CA_1      FOODS           0
3      CA_2    HOBBIES           0
4      CA_2  HOUSEHOLD           0
5      CA_2      FOODS          19
6      CA_3    HOBBIES           0
7      CA_3  HOUSEHOLD           0
8      CA_3      FOODS          12
9      CA_4    HOBBIES           0
10     CA_4  HOUSEHOLD           0
11     CA_4      FOODS           0
12     TX_1    HOBBIES           0
13     TX_1  HOUSEHOLD           0
14     TX_1      FOODS           2
15     TX_2    HOBBIES           0
16     TX_2  HOUSEHOLD           0
17     TX_2      FOODS          18
18     TX_3    HOBBIES           0
19     TX_3  HOUSEHOLD           0
20     TX_3      FOODS          14
21     WI_1    HOBBIES           0
22     WI_1  HOUSEHOLD           1
23     WI_1      FOODS           2
24     WI_2    HOBBIES           0
25     WI_2  HOUSEHOLD           0
26     WI_2      FOODS           7
27     WI_3    HOBBIES           0
28     WI_3  HOUSEHOLD           0
29     WI_3      FOODS           3

In [30]:
data['store_id'] = data.store_id.astype('str')
data['cat_id'] = data.cat_id.astype('str')

In [31]:
data['store_cat_id']=data['store_id']+'_'+data['cat_id']
data



store_id     cat_id  Total_sale    store_cat_id
0      CA_1    HOBBIES           0    CA_1_HOBBIES
1      CA_1  HOUSEHOLD           0  CA_1_HOUSEHOLD
2      CA_1      FOODS           0      CA_1_FOODS
3      CA_2    HOBBIES           0    CA_2_HOBBIES
4      CA_2  HOUSEHOLD           0  CA_2_HOUSEHOLD
5      CA_2      FOODS          19      CA_2_FOODS
6      CA_3    HOBBIES           0    CA_3_HOBBIES
7      CA_3  HOUSEHOLD           0  CA_3_HOUSEHOLD
8      CA_3      FOODS          12      CA_3_FOODS
9      CA_4    HOBBIES           0    CA_4_HOBBIES
10     CA_4  HOUSEHOLD           0  CA_4_HOUSEHOLD
11     CA_4      FOODS           0      CA_4_FOODS
12     TX_1    HOBBIES           0    TX_1_HOBBIES
13     TX_1  HOUSEHOLD           0  TX_1_HOUSEHOLD
14     TX_1      FOODS           2      TX_1_FOODS
15     TX_2    HOBBIES           0    TX_2_HOBBIES
16     TX_2  HOUSEHOLD           0  TX_2_HOUSEHOLD
17     TX_2      FOODS          18      TX_2_FOODS
18     TX_3    HOBBIES           0    TX_3_HOBBIES
19     TX_3  HOUSEHOLD           0  TX_3_HOUSEHOLD
20     TX_3      FOODS          14      TX_3_FOODS
21     WI_1    HOBBIES           0    WI_1_HOBBIES
22     WI_1  HOUSEHOLD           1  WI_1_HOUSEHOLD
23     WI_1      FOODS           2      WI_1_FOODS
24     WI_2    HOBBIES           0    WI_2_HOBBIES
25     WI_2  HOUSEHOLD           0  WI_2_HOUSEHOLD
26     WI_2      FOODS           7      WI_2_FOODS
27     WI_3    HOBBIES           0    WI_3_HOBBIES
28     WI_3  HOUSEHOLD           0  WI_3_HOUSEHOLD
29     WI_3      FOODS           3      WI_3_FOODS

In [32]:
px.bar(data, x="store_cat_id", y="Total_sale", color="store_cat_id", title="Total Sale base on Christmas in each Store ",text='Total_sale')

### 1.2.2  mean sale in every state

In [33]:
# here i am grouping data by days and takeing mean 
sales_mean=df_final.groupby(['state_id','d'],as_index=False,sort=False).agg(mean_sold=('sold',np.mean))
sales_mean

state_id       d  mean_sold
0          CA     d_1   1.163906
1          TX     d_1   1.031814
2          WI     d_1   0.983711
3          CA     d_2   1.131929
4          TX     d_2   1.052804
...       ...     ...        ...
5818       TX  d_1940   1.495682
5819       WI  d_1940   1.610801
5820       CA  d_1941   2.020663
5821       TX  d_1941   1.619657
5822       WI  d_1941   1.626654

[5823 rows x 3 columns]

In [34]:
kk=sales_mean.groupby('state_id')
ff=kk.groups
CA=sales_mean.iloc[ff['CA']].reset_index()
TX=sales_mean.iloc[ff['TX']].reset_index()
WI=sales_mean.iloc[ff['WI']].reset_index()
TX=TX.drop(columns='index')
CA=CA.drop(columns='index')
WI=WI.drop(columns='index')

In [36]:
D={'d':(CA['d']),'CA':CA['mean_sold'],'TX':TX['mean_sold'],'WI':WI['mean_sold']}
data=pd.DataFrame(D)

In [37]:
data

d        CA        TX        WI
0        d_1  1.163906  1.031814  0.983711
1        d_2  1.131929  1.052804  0.908932
2        d_3  0.828796  0.741008  0.754018
3        d_4  0.905789  0.806931  0.763529
4        d_5  0.813791  0.646332  0.361758
...      ...       ...       ...       ...
1936  d_1937  1.285503  1.134252  1.207281
1937  d_1938  1.336258  1.001640  1.257680
1938  d_1939  1.429157  1.345031  1.401443
1939  d_1940  1.894310  1.495682  1.610801
1940  d_1941  2.020663  1.619657  1.626654

[1941 rows x 4 columns]

In [38]:

# Time Series With Range Slider
import plotly.express as px
import pandas as pd

fig = px.line(data,x='d', y=data.columns, title='states VS mean sale')
fig.update_xaxes(rangeslider_visible=True)
fig.update_xaxes(title_text='days')
fig.update_yaxes(title_text='mean_sold')
fig.show()

### Observation:
1. all states mean sale are almost overlap on each other. cannot found any difference.
2. but still there is zero sale present in California and Wisconsin ,but in Taxas we didnt found any zero sale.  
3. initially  sales increases and it suddenly decreases at the end of the year. right now i cant say this confidently as plot is totally overlap . we have to dig deep into it.

### 1.2.3  mean sales in stores

In [39]:
df_final[(df_final["state_id"]=="CA") & (df_final['d']=='d_1')]['sold'].mean()

1.1639061987536898

In [40]:
# lets see how many store are there in in each state
df_final.groupby(['state_id'],as_index=False,sort=False).agg(unique_store=('store_id',pd.Series.unique))


state_id              unique_store
0       CA  [CA_1, CA_2, CA_3, CA_4]
1       TX        [TX_1, TX_2, TX_3]
2       WI        [WI_1, WI_2, WI_3]

* there are 10 stores in total in which 4 store belong to california , 3 store belong to Taxas and 3 store wisconsin.let explore each store one by one.

In [41]:

sales_mean=df_final.groupby(['store_id','d'],as_index=False,sort=False).agg(mean_sold=('sold',np.mean))

In [42]:
# preparing data 
kk=sales_mean.groupby('store_id')
ff=kk.groups 
CA_1=sales_mean.iloc[ff['CA_1']].reset_index()
CA_2=sales_mean.iloc[ff['CA_2']].reset_index()
CA_3=sales_mean.iloc[ff['CA_3']].reset_index()
CA_4=sales_mean.iloc[ff['CA_4']].reset_index()

TX_1=sales_mean.iloc[ff['TX_1']].reset_index()
TX_2=sales_mean.iloc[ff['TX_2']].reset_index()
TX_3=sales_mean.iloc[ff['TX_3']].reset_index()

WI_1=sales_mean.iloc[ff['WI_1']].reset_index()
WI_2=sales_mean.iloc[ff['WI_2']].reset_index()
WI_3=sales_mean.iloc[ff['WI_3']].reset_index()

CA_1=CA_1.drop(columns='index')
CA_2=CA_2.drop(columns='index')
CA_3=CA_3.drop(columns='index')
CA_4=CA_4.drop(columns='index')

TX_1=TX_1.drop(columns='index')
TX_2=TX_2.drop(columns='index')
TX_3=TX_3.drop(columns='index')

WI_1=WI_1.drop(columns='index')
WI_2=WI_2.drop(columns='index')
WI_3=WI_3.drop(columns='index')

In [43]:
print(CA_1)

     store_id       d  mean_sold
0        CA_1     d_1   1.422434
1        CA_1     d_2   1.362742
2        CA_1     d_3   0.923582
3        CA_1     d_4   1.000656
4        CA_1     d_5   0.862578
...       ...     ...        ...
1936     CA_1  d_1937   1.310266
1937     CA_1  d_1938   1.356510
1938     CA_1  d_1939   1.453919
1939     CA_1  d_1940   1.890456
1940     CA_1  d_1941   2.062643

[1941 rows x 3 columns]


### 1.2.3.1 mean sales in California:

In [44]:
D={'d':(CA_1['d']),'CA_1':CA_1['mean_sold'],'CA_2':CA_2['mean_sold'],'CA_3':CA_3['mean_sold'],'CA_4':CA_4['mean_sold']}
data=pd.DataFrame(D)

In [45]:

import plotly.express as px
import pandas as pd

fig = px.line(data,x='d', y=data.columns, title='california VS mean sale')
fig.update_xaxes(rangeslider_visible=True)
fig.update_xaxes(title_text='days')
fig.update_yaxes(title_text='mean_sold')
fig.show()

### 1.2.3.2 mean sales in Taxas:

In [46]:
D={'d':(TX_1['d']),'TX_1':TX_1['mean_sold'],'TX_2':TX_2['mean_sold'],'TX_3':TX_3['mean_sold']}
data=pd.DataFrame(D)

In [47]:
data

d      TX_1      TX_2      TX_3
0        d_1  0.838308  1.263365  0.993768
1        d_2  0.881273  1.291243  0.985897
2        d_3  0.597573  0.895704  0.729747
3        d_4  0.740571  0.968842  0.711381
4        d_5  0.555592  0.817317  0.566087
...      ...       ...       ...       ...
1936  d_1937  1.032142  1.153821  1.216792
1937  d_1938  0.901279  1.025254  1.078386
1938  d_1939  1.201705  1.393572  1.439816
1939  d_1940  1.366678  1.574943  1.545425
1940  d_1941  1.516563  1.711053  1.631355

[1941 rows x 4 columns]

In [48]:
import plotly.express as px
import pandas as pd

fig = px.line(data,x='d', y=data.columns, title='california VS mean sale')
fig.update_xaxes(rangeslider_visible=True)
fig.update_xaxes(title_text='days')
fig.update_yaxes(title_text='mean_sold')
fig.show()

### 1.2.3.3 mean sales in Wisconsin:

In [49]:
D={'d':(WI_1['d']),'WI_1':WI_1['mean_sold'],'WI_2':WI_2['mean_sold'],'WI_3':WI_3['mean_sold']}
data=pd.DataFrame(D)

In [50]:

import plotly.express as px
import pandas as pd

fig = px.line(data,x='d', y=data.columns, title='california VS mean sale')
fig.update_xaxes(rangeslider_visible=True)
fig.update_xaxes(title_text='days')
fig.update_yaxes(title_text='mean_sold')
fig.show()

In [52]:
for i in range(1,7):
    print(365*i)

365
730
1095
1460
1825
2190


### Observation :
* as all the plots are more or less look similar.all the observation which are written below are applicable to all plots.
1. cant found any good information as there is no perodicity in sale. sale is erotic in neature.
2. but yes there is one info i found in this plot.eroticity is present in months or in years but not in days.i mean we can clearly see the trend for short period of intervals.it could be last 2 days or last 3 days and so on.

## 1.3.1 Lets see sales in state wise and store wise using barplot.

In [53]:
df_final.head()

id        item_id    dept_id   cat_id store_id  \
0  HOBBIES_1_001_CA_1_evaluation  HOBBIES_1_001  HOBBIES_1  HOBBIES     CA_1   
1  HOBBIES_1_002_CA_1_evaluation  HOBBIES_1_002  HOBBIES_1  HOBBIES     CA_1   
2  HOBBIES_1_003_CA_1_evaluation  HOBBIES_1_003  HOBBIES_1  HOBBIES     CA_1   
3  HOBBIES_1_004_CA_1_evaluation  HOBBIES_1_004  HOBBIES_1  HOBBIES     CA_1   
4  HOBBIES_1_005_CA_1_evaluation  HOBBIES_1_005  HOBBIES_1  HOBBIES     CA_1   

  state_id    d  sold        date  wm_yr_wk  ... month  year  event_name_1  \
0       CA  d_1     0  2011-01-29     11101  ...     1  2011      No_event   
1       CA  d_1     0  2011-01-29     11101  ...     1  2011      No_event   
2       CA  d_1     0  2011-01-29     11101  ...     1  2011      No_event   
3       CA  d_1     0  2011-01-29     11101  ...     1  2011      No_event   
4       CA  d_1     0  2011-01-29     11101  ...     1  2011      No_event   

   event_type_1 event_name_2 event_type_2 snap_CA snap_TX  snap_WI  sell_price  
0      No_event     No_event     No_event       0       0        0         NaN  
1      No_event     No_event     No_event       0       0        0         NaN  
2      No_event     No_event     No_event       0       0        0         NaN  
3      No_event     No_event     No_event       0       0        0         NaN  
4      No_event     No_event     No_event       0       0        0         NaN  

[5 rows x 22 columns]

In [54]:
df_final.columns

Index(['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id', 'd',
       'sold', 'date', 'wm_yr_wk', 'weekday', 'wday', 'month', 'year',
       'event_name_1', 'event_type_1', 'event_name_2', 'event_type_2',
       'snap_CA', 'snap_TX', 'snap_WI', 'sell_price'],
      dtype='object')

### 1.3.1.1  Total Sales in each state.

In [55]:
data=df_final.groupby('state_id',as_index=False,sort=False).agg(Total_sale=('sold',np.sum))
data

state_id  Total_sale
0       CA    29196717
1       TX    19228405
2       WI    18502051

In [58]:
# # https://plotly.com/python/bar-charts/
import plotly.graph_objects as go

x = data['state_id']
y = data['Total_sale']

fig = go.Figure(data=[go.Bar(
             x=x, y=y,
             text=y,
             textposition='auto',marker_color='blue'
         )])
fig.update_layout(title_text='total product sale in each state')

fig.show()

In [ ]:
px.bar(data, x="state_id", y="Total_sale", color="state_id", title="Stores vs Mean sales",text='Total_sale')

### Observation:
1. california is largest state from USA in term of population and having population 37,999,878 ,Taxas is second largest state from USA in terms of population  and having population 26,060,796 and wisconsin having 5,724,554 population.
2. so yes total sales in california is high but population is also hign in california.surprisingly wisconsin having less population  as compare to taxas but still they have similar total sale. 

### 1.3.1.2  Total Sales in each Store.

In [59]:
data=df_final.groupby('store_id',as_index=False,sort=False).agg(Total_sale=('sold',np.sum))
data

store_id  Total_sale
0     CA_1     7832248
1     CA_2     5818395
2     CA_3    11363540
3     CA_4     4182534
4     TX_1     5692823
5     TX_2     7329642
6     TX_3     6205940
7     WI_1     5261506
8     WI_2     6697988
9     WI_3     6542557

In [60]:
px.bar(data, x="store_id", y="Total_sale", color="store_id", title="Total sale in each store",text='Total_sale')

### Observation:
1. california store CA_3 have higest total sale amoungest all. and in california sales also not stable.

### 1.4.1Total sales based on product categories  across all three state .

In [61]:
data=df_final.groupby('cat_id',as_index=False,sort=False).agg(Total_sale=('sold',np.sum))
data

cat_id  Total_sale
0    HOBBIES     6240656
1  HOUSEHOLD    14764090
2      FOODS    45922427

In [62]:
px.bar(data, x="cat_id", y="Total_sale", color="cat_id", title="Total sale in USA based on each category",text='Total_sale')

### Observation :
1. sales of product belong to Foods category is way higher than remaining category.

### 1.4.2 sales based on product category in every state.

In [63]:
data=df_final.groupby(['state_id','cat_id'],as_index=False,sort=False).agg(Total_sale=('sold',np.sum))
data

state_id     cat_id  Total_sale
0       CA    HOBBIES     3095587
1       CA  HOUSEHOLD     6565267
2       CA      FOODS    19535863
3       TX    HOBBIES     1624130
4       TX  HOUSEHOLD     4432169
5       TX      FOODS    13172106
6       WI    HOBBIES     1520939
7       WI  HOUSEHOLD     3766654
8       WI      FOODS    13214458

In [64]:
data['state_id'] = data.state_id.astype('str')
data['cat_id'] = data.cat_id.astype('str')


In [65]:
data['state_cat_id']=data['state_id']+'_'+data['cat_id']


In [66]:
data

state_id     cat_id  Total_sale  state_cat_id
0       CA    HOBBIES     3095587    CA_HOBBIES
1       CA  HOUSEHOLD     6565267  CA_HOUSEHOLD
2       CA      FOODS    19535863      CA_FOODS
3       TX    HOBBIES     1624130    TX_HOBBIES
4       TX  HOUSEHOLD     4432169  TX_HOUSEHOLD
5       TX      FOODS    13172106      TX_FOODS
6       WI    HOBBIES     1520939    WI_HOBBIES
7       WI  HOUSEHOLD     3766654  WI_HOUSEHOLD
8       WI      FOODS    13214458      WI_FOODS

In [67]:
px.bar(data, x="state_cat_id", y="Total_sale", color="state_cat_id", title="Total sale in each state based on category",text='Total_sale')

### Observation:
1. in every state food category have higher sale,then secondly Household category have higher sale.

###  1.5.1 day sale in USA (CA,TX,WI)

In [68]:
df_final.columns

Index(['id', 'item_id', 'dept_id', 'cat_id', 'store_id', 'state_id', 'd',
       'sold', 'date', 'wm_yr_wk', 'weekday', 'wday', 'month', 'year',
       'event_name_1', 'event_type_1', 'event_name_2', 'event_type_2',
       'snap_CA', 'snap_TX', 'snap_WI', 'sell_price'],
      dtype='object')

In [69]:
df_final[['date','weekday','wday','month','year']]

date   weekday  wday  month  year
0         2011-01-29  Saturday     1      1  2011
1         2011-01-29  Saturday     1      1  2011
2         2011-01-29  Saturday     1      1  2011
3         2011-01-29  Saturday     1      1  2011
4         2011-01-29  Saturday     1      1  2011
...              ...       ...   ...    ...   ...
59181085  2016-05-22    Sunday     2      5  2016
59181086  2016-05-22    Sunday     2      5  2016
59181087  2016-05-22    Sunday     2      5  2016
59181088  2016-05-22    Sunday     2      5  2016
59181089  2016-05-22    Sunday     2      5  2016

[59181090 rows x 5 columns]

In [70]:
data=df_final.groupby(['weekday'],as_index=False,sort=False).agg(Total_sale=('sold',np.sum))
data

weekday  Total_sale
0   Saturday    11594343
1     Sunday    11482526
2     Monday     9139596
3    Tuesday     8448634
4  Wednesday     8346126
5   Thursday     8398928
6     Friday     9517020

In [71]:
px.bar(data, x="weekday", y="Total_sale", color="weekday", title="Total sale day wise",text='Total_sale')

### Observation:
1. every monday sales start to fall down till thursday and again start to go up/

### 1.5.2 Sales based on month

In [72]:
data=df_final.groupby(['month'],as_index=False,sort=False).agg(Total_sale=('sold',np.sum))
data

month  Total_sale
0       1     5345511
1       2     5889589
2       3     6387210
3       4     6197793
4       5     6033079
5       6     5250212
6       7     5443975
7       8     5571730
8       9     5321889
9      10     5380589
10     11     4993624
11     12     5111972

In [73]:
data['month_name']=['January','February','March','April','May','June','July','August','September','Octobar','November','December']

In [74]:
data

month  Total_sale month_name
0       1     5345511    January
1       2     5889589   February
2       3     6387210      March
3       4     6197793      April
4       5     6033079        May
5       6     5250212       June
6       7     5443975       July
7       8     5571730     August
8       9     5321889  September
9      10     5380589    Octobar
10     11     4993624   November
11     12     5111972   December

In [75]:
px.bar(data, x="month_name", y="Total_sale", color="month_name", title="Total sale month wise",text='Total_sale')

### Observation:
1. from february sales stat increases and it slowly increases up to march then it start decreasing untill june.
2. in march highest sales happen. and in novembar lowest sales happen.  

### 1.5.3 Sales Based on Event_type_1

In [76]:
df_final['event_name_1'].unique()

array(['No_event', 'SuperBowl', 'ValentinesDay', 'PresidentsDay',
       'LentStart', 'LentWeek2', 'StPatricksDay', 'Purim End',
       'OrthodoxEaster', 'Pesach End', 'Cinco De Mayo', "Mother's day",
       'MemorialDay', 'NBAFinalsStart', 'NBAFinalsEnd', "Father's day",
       'IndependenceDay', 'Ramadan starts', 'Eid al-Fitr', 'LaborDay',
       'ColumbusDay', 'Halloween', 'EidAlAdha', 'VeteransDay',
       'Thanksgiving', 'Christmas', 'Chanukah End', 'NewYear',
       'OrthodoxChristmas', 'MartinLutherKingDay', 'Easter'], dtype=object)

In [ ]:
# 'df_final.head()

In [77]:
data=df_final.groupby(['event_type_1'],as_index=False).agg(Mean_sale=('sold',np.mean))
data

event_type_1  Mean_sale
0     Cultural   1.138688
1     National   0.966170
2     No_event   1.135105
3    Religious   1.129099
4     Sporting   1.174026

In [78]:
px.bar(data, x="event_type_1", y="Mean_sale", color="event_type_1", title="Mean sale based on event_type1",text='Mean_sale')

### Observation:
1. mean_sale is higher on soprting event as compare to other event.  
2. day having national event have lower mean_sale as compare to other event.
3. so i can say that there is an impact of Event on sales.

### 1.5.4 Mean sale based on event_type_2

In [79]:
data=df_final.groupby(['event_type_2'],as_index=False).agg(Mean_sale=('sold',np.mean))
data

event_type_2  Mean_sale
0     Cultural   1.199464
1     No_event   1.130706
2    Religious   1.276943

In [80]:
px.bar(data, x="event_type_2", y="Mean_sale", color="event_type_2", title="Mean sale based on event_type2",text='Mean_sale')

### Observation :
1. there is impact of Religions and Cultural events on sale. sale is high. 

### 1.5.5 Total sale when there is snap and no snap

In [81]:
data1=df_final[df_final['state_id']=='CA'][['sold','snap_CA']]
data1

sold  snap_CA
0            0        0
1            0        0
2            0        0
3            0        0
4            0        0
...        ...      ...
59162791     0        0
59162792     0        0
59162793     0        0
59162794     4        0
59162795     0        0

[23672436 rows x 2 columns]

In [82]:
data1=data1.groupby(['snap_CA'],as_index=False).agg(Total_sale=('sold',np.sum))
data1

snap_CA  Total_sale
0        0    19069726
1        1    10126991

In [83]:
data2=df_final[df_final['state_id']=='TX'][['sold','snap_TX']]
data2

sold  snap_TX
12196        0        0
12197        0        0
12198        0        0
12199        1        0
12200        0        0
...        ...      ...
59171938     1        0
59171939     0        0
59171940     3        0
59171941     2        0
59171942     2        0

[17754327 rows x 2 columns]

In [84]:
data2=data2.groupby(['snap_TX'],as_index=False).agg(Total_sale=('sold',np.sum))
data2

snap_TX  Total_sale
0        0    12416557
1        1     6811848

In [85]:
data3=df_final[df_final['state_id']=='WI'][['sold','snap_WI']]
data3

sold  snap_WI
21343        0        0
21344        0        0
21345        0        0
21346        2        0
21347        0        0
...        ...      ...
59181085     1        0
59181086     0        0
59181087     2        0
59181088     0        0
59181089     1        0

[17754327 rows x 2 columns]

In [86]:
data3=data3.groupby(['snap_WI'],as_index=False).agg(Total_sale=('sold',np.sum))
data3

snap_WI  Total_sale
0        0    11569147
1        1     6932904

In [87]:
data1['snap_CA'][0]

0

In [88]:
import plotly.graph_objects as go
dataset_name=["CA","TX","WI"]
d1 = [data1['Total_sale'][0], data2['Total_sale'][0], data3['Total_sale'][0]]
d2 = [data1['Total_sale'][1], data2['Total_sale'][1],data3['Total_sale'][1]]
fig = go.Figure(data=[
    go.Bar(name='No_snap', x=dataset_name, y=d2,text=d2,textposition='auto'),
    go.Bar(name='snap', x=dataset_name, y=d1,text=d1,textposition='auto')
])
# Change the bar mode
fig.update_xaxes(title_text='states')
fig.update_yaxes(title_text='Total sum')
fig.update_layout(barmode='group')
fig.update_layout(title_text='Total sale based on Snap days and Non Snap say')
fig.show()

### Observation:
1. Sales on all the states during the Snap days are high than Non-Snap days.

## Overall conclusion:
1. Using Dawncasting we can reduce the dataFrame size.
2. every year in USA(CA,TX,WI) mean sale per day increases.
3. there are some days where sale was zero. on every Christmas, sale was zero.only some food sale happen.
4. at the start of every year sale start to increase and after march it slowly start to decrease.
5. even in weekends sales start to increases and after sunday it start to decrease till friday.
6. california having large sale but it bocz of large population.
7. Food category product have highest sale as compare to other.
8. day having sporting event and religious have good impact on sale.sale is high on this events.
9. day having snap result in high sales.